### PR TITLE
IC-913 AWS SNS integration for publishing messages to a shared event queue

### PR DESCRIPTION
**Pull request by tomsmyers**
## What does this pull request do?

publishes an event to an SNS topic when a referral is sent.

## What is the intent behind these changes?

allow other services to pick up these events as they see fit

### CODE REVIEW

File: docker-compose-localstack.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  localstack:
+    container_name: "${LOCALSTACK_DOCKER_NAME-localstack_main}"
+    image: localstack/localstack
+    network_mode: bridge
+    ports:
+      - "4566:4566"
+      - "4571:4571"
+    environment:
+      - LOCALSTACK_SERVICES=${LOCALSTACK_SERVICES-sns}
+      - LOCALSTACK_DEFAULT_REGION=${LOCALSTACK_DEFAUL_REGION-eu-west-1}
```
Comment (sldblog) [inline docker-compose-localstack.yml:13]: 🙋‍♂️ I'm not sure of the implications, but we default to `eu-west-2` (London) due to data handling related to individuals

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/SNSConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.sns.SnsAsyncClient
+import java.net.URI
+
+@Configuration
+class SNSConfiguration(
+  @Value("\${aws.region}") private val region: String,
+  @Value("\${aws.access-key-id}") private val accessKeyId: String,
+  @Value("\${aws.secret-access-key}") private val secretAccessKey: String,
+  @Value("\${aws.sns.uri}") private val uri: String,
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/SNSConfiguration.kt:16]: It is technically a URI but it's probably more colloquial to define an [ARN](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html).
What do you think about using `aws.sns.arn`?

🙋‍♂️ The env variable currently is `AWS_SNS_TOPIC_ARN` which doesn't seem to match up to this, unless I'm missing something.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/SNSConfiguration.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.sns.SnsAsyncClient
+import java.net.URI
+
+@Configuration
+class SNSConfiguration(
+  @Value("\${aws.region}") private val region: String,
+  @Value("\${aws.access-key-id}") private val accessKeyId: String,
+  @Value("\${aws.secret-access-key}") private val secretAccessKey: String,
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/SNSConfiguration.kt:15]: _prefix_
I think we need specific access/secrets for the SNS integration. Is it possible to prefix them like that?

I.e. something like
```
aws.sns.region
aws.sns.access_key_id
aws.sns.secret_access_key
```

File: README.md [hunk 4]
```diff
@@ -28,6 +28,15 @@ docker-compose up -d
 SPRING_PROFILES_ACTIVE=local ./gradlew bootRun
 ```
 
+### Running with AWS LocalStack for SNS (Simple Notification Service)
+
+- Install the AWS CLI with Homebrew: `brew install awscli`
+- Configure the CLI for localstack: `aws configure --profile localstack` and set `AWS Access Key ID = test`, `AWS Secret Access Key = test`, `Default region name = eu-west-1`, `Default output format = json`
+- Run localstack with docker-compose: `docker-compose -f docker-compose-localstack.yml`
+- Create an SNS topic: `aws --profile localstack --endpoint-url=http://localhost:4566 sns create-topic --name localstack-topic`
```
Comment (sldblog) [inline README.md:36]: Using a specific topic name might help if someone has many SNS apps :)
```suggestion
- Create an SNS topic: `aws --profile localstack --endpoint-url=http://localhost:4566 sns create-topic --name intervention-events-local`
```

File: README.md [hunk 5]
```diff
@@ -28,6 +28,15 @@ docker-compose up -d
 SPRING_PROFILES_ACTIVE=local ./gradlew bootRun
 ```
 
+### Running with AWS LocalStack for SNS (Simple Notification Service)
+
+- Install the AWS CLI with Homebrew: `brew install awscli`
+- Configure the CLI for localstack: `aws configure --profile localstack` and set `AWS Access Key ID = test`, `AWS Secret Access Key = test`, `Default region name = eu-west-1`, `Default output format = json`
```
Comment (sldblog) [inline README.md:34]: It's a minor thing, but we won't use regions outside the UK, so for consistency, it's probably better to use
```suggestion
- Configure the CLI for localstack: `aws configure --profile localstack` and set `AWS Access Key ID = test`, `AWS Secret Access Key = test`, `Default region name = eu-west-2`, `Default output format = json`
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/NotifyListeners.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+@Component
+class NotifyReferralEventListener(
+  @Value("\${interventions-ui.baseurl}") private val interventionsUIBaseURL: String,
+  @Value("\${interventions-ui.locations.sent-referral}") private val interventionsUISentReferralLocation: String,
+  private val notifyService: NotifyService,
+) : ApplicationListener<ReferralEvent> {
+  override fun onApplicationEvent(event: ReferralEvent) {
+    when (event.type) {
+      ReferralEventType.SENT -> {
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/NotifyListeners.kt:18]: _musing, not blocking_
Checking: this is part of #66, right? (This PR could have a "base" on top of that branch, so it only shows additions, with the trade-off of some extra rebases -- or it could be split to "foundations" and 2 PRs adding listener logic?)

**Review by sldblog — COMMENTED** [review]
Nice. Most comments are about naming/prefixes/regions

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/NotifyListeners.kt (hunk 6 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/NotifyListeners.kt:18]: this PR contains all the commits from the other one - the most recent commit contains just the new changes.

not sure how to isolate this. you can create a PR on the other branch, but that isn't want we want, it needs to go into main. so as soon as the other branch is merged this will become clearer.

File: README.md (hunk 5 above)
Comment (tomsmyers) [reply README.md:34]: sure thing

File: README.md (hunk 4 above)
Comment (tomsmyers) [reply README.md:36]: good idea.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/SNSConfiguration.kt (hunk 3 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/SNSConfiguration.kt:15]: sure thing

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/SNSConfiguration.kt (hunk 2 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/SNSConfiguration.kt:16]: uri is different to arn. the uri is the endpoint i.e. localhost:5466 and arn is the topic definition. i'll rename uri to be clearer.

File: docker-compose-localstack.yml (hunk 1 above)
Comment (tomsmyers) [reply docker-compose-localstack.yml:13]: no implications for localstack, but i'll change for clarity.

File: docker-compose-localstack.yml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  localstack:
+    container_name: "${LOCALSTACK_DOCKER_NAME-localstack_main}"
+    image: localstack/localstack
+    network_mode: bridge
+    ports:
+      - "4566:4566"
+      - "4571:4571"
+    environment:
+      - LOCALSTACK_SERVICES=${LOCALSTACK_SERVICES-sns}
+      - LOCALSTACK_DEFAULT_REGION=${LOCALSTACK_DEFAUL_REGION-eu-west-2}
```
Comment (karenmillermoj) [inline docker-compose-localstack.yml:13]: Looks like a typo? {LOCALSTACK_DEFAUL_REGION-eu-west-2}  - missing T

Comment (tomsmyers) [reply docker-compose-localstack.yml]: good spot!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.context.ApplicationListener
+import org.springframework.stereotype.Component
+import org.springframework.web.util.UriComponentsBuilder
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.NotifyService
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.SNSService
+
+@Component
+class ReferralEventListener(
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt:12]: i've consolidated all the event side effects here so it's easier to see what is being triggered.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  override fun onApplicationEvent(event: ReferralEvent) {
+    when (event.type) {
+      ReferralEventType.SENT -> {
+        sendReferralSentEmail(event)
+        publishReferralSentToSNSTopic(event)
+      }
+    }
+  }
+
+  private fun sendReferralSentEmail(event: ReferralEvent) {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt:28]: so this method is just copied code

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    val location = UriComponentsBuilder.fromHttpUrl(interventionsUIBaseURL)
+      .path(interventionsUISentReferralLocation)
+      .buildAndExpand(event.referral.id)
+      .toUri()
+
+    // fixme: this email address will eventually come from the provider associated with the referral
+    notifyService.referralSent(event.referral.referenceNumber!!, location, "tom.myers@digital.justice.gov.uk")
+  }
+
+  private fun publishReferralSentToSNSTopic(event: ReferralEvent) {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt:38]: but this is new code

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListenerTest.kt [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+          id = UUID.fromString("68df9f6c-3fcb-4ec6-8fcf-96551cd9b080"),
+          referenceNumber = "HAS71263",
+          serviceUserCRN = "X123456"
+        )
+      )
+    )
+  }
+
+  @Test
+  fun `referral sent event calls notify service with valid url`() {
```
Comment (tomsmyers) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListenerTest.kt:43]: as above - this code is unchanged, it's just moved

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListenerTest.kt [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    simulateEvent()
+
+    val locationCaptor = argumentCaptor<URI>()
+    verify(notifyService).referralSent(eq("HAS71263"), locationCaptor.capture(), any())
+
+    assertThat(locationCaptor.firstValue.toString()).isEqualTo("http://example.com/referral/68df9f6c-3fcb-4ec6-8fcf-96551cd9b080")
+  }
+
+  @Test
+  fun `referral sent event calls sns service with valid json`() {
```
Comment (tomsmyers) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListenerTest.kt:53]: but this is new

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  private val notifyService: NotifyService,
+  private val snsService: SNSService,
+  private val objectMapper: ObjectMapper,
+  @Value("\${interventions-ui.baseurl}") private val interventionsUIBaseURL: String,
+  @Value("\${interventions-ui.locations.sent-referral}") private val interventionsUISentReferralLocation: String,
+) : ApplicationListener<ReferralEvent> {
+  override fun onApplicationEvent(event: ReferralEvent) {
+    when (event.type) {
+      ReferralEventType.SENT -> {
+        sendReferralSentEmail(event)
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt:22]: make error handling more explicit so all these run all the time

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/NotifyService.kt [hunk 14]
```diff
@@ -13,21 +13,23 @@ class NotifyService(
...
-        // fixme: this failure is super important and we need a better way to reason about async errors of this nature
-        log.error("referral sent notification failed", e)
-      }
+    if (!enabled) {
+      return
+    }
+
+    try {
+      client.sendEmail(
+        referralSentTemplateID,
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/NotifyService.kt:22]: Given we couple the configuration of this class to be a notification about a sent referral, we could also make it a listener to sent referral. Not sure if that makes things worse or better 🤷‍♂️ 
```kt
-class NotifyService(...)
+class NotifyProviderWithReferral(...) : ApplicationListener<ReferralEvent>
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  private fun publishReferralSentToSNSTopic(event: ReferralEvent) {
+    // fixme: what data needs to go in here
+    val referralSentMessage = mapOf(
+      "type" to "referral-sent",
+      "description" to "A referral has been sent to a Service Provider",
+      "id" to event.referral.id,
+    )
+    snsService.publish(objectMapper.writeValueAsString(referralSentMessage))
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt:45]: I think the `fixme` comment highlights that this class might prove difficult to test:

- there's logic about location URI building for emails that need testing
- there's logic about constructing a referral SNS message that needs testing

Those tests don't seem related.

File: README.md [hunk 16]
```diff
@@ -28,6 +28,15 @@ docker-compose up -d
 SPRING_PROFILES_ACTIVE=local ./gradlew bootRun
 ```
 
+### Running with AWS LocalStack for SNS (Simple Notification Service)
```
Comment (sldblog) [inline README.md:31]: ❤️ these instructions

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/SNSConfiguration.kt [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+class SNSConfiguration(
+  @Value("\${aws.sns.provider}") private val provider: String,
+  @Value("\${aws.sns.region}") private val region: String,
+  @Value("\${aws.sns.access-key-id}") private val accessKeyId: String,
+  @Value("\${aws.sns.secret-access-key}") private val secretAccessKey: String,
+  @Value("\${aws.sns.endpoint.uri:}") private val uri: String,
+
+) {
+  @Bean
+  fun snsClient(): SnsAsyncClient {
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/SNSConfiguration.kt:21]: Would it help with tests down the line if we use the non-async [SnsClient](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/sns/SnsClient.html)?

It seems to respond with a `PublishResponse` directly, which sounds OK since we're already firing this in an async listener.

**Review by sldblog — COMMENTED** [review]
👍 Looking good

Capturing some of the conversations + new thoughts

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/SNSConfiguration.kt (hunk 17 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/SNSConfiguration.kt:21]: good idea.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt (hunk 15 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ReferralEventListener.kt:45]: agreed - changed.

**Review by sldblog — APPROVED** [review]
👍 Looks good

Full disclosure: I haven't had an opportunity to follow the readme steps
